### PR TITLE
Refactor: collection page tests

### PR DESCRIPTION
**Pull request by alexanderleegs**
This PR introduces tests to the newly refactored collection page endpoints. See [this document](https://docs.google.com/document/d/1uGzg0XA08xvyfvBF2e_2giNXgO1micmLY6Cz9bEq3Qo/edit?pli=1#heading=h.3eu9rwalz5km) for a full discussion on the structure of the backend testing.

We have elected to use unit testing for each of the backend components, since end-to-end tests are covered by the isomercms frontend. We have used the `supertest` library to help with the mocking of requests in testing.

Of special note is the router layer unit test for collection pages - I faced a bug I was unable to resolve when using the `getRouter` method of the class, where requests were directed to the correct method but the path params could not be parsed. We have verified that this bug only exists in testing, and the current method of using the router instead of attaching each route explicitly works in practice as well as in end to end testing. The workaround for this method has been to use a different implementation in testing by attaching the routes directly (i.e. `app.post(
    "/:siteName/collections/:collectionName/pages"` instead of `app.use(collectionPagesRouter.getRouter())`), but if anyone has a better method of doing this, please let me know.

### CODE REVIEW

File: services/fileServices/MdPageServices/__tests__/collectionPageService.spec.js [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+        directoryName,
+        fileContent: mockMarkdownContent,
+        sha: oldSha,
+      })
+    })
+  })
+
+  describe("Delete", () => {
+    it("Deleting pages works correctly", async () => {
+      await expect(
```
Comment (prestonlimlianjie) [inline services/fileServices/MdPageServices/__tests__/collectionPageService.spec.js:170]: Should also expect `mockCollectionYmlService.deleteItemFromorder` to have been called

File: services/fileServices/MdPageServices/__tests__/subcollectionPageService.spec.js [hunk 2]
```diff
@@ -0,0 +1,238 @@
...
+        fileContent: mockMarkdownContent,
+        sha: oldSha,
+      })
+    })
+  })
+
+  describe("Delete", () => {
+    it("Deleting pages works correctly", async () => {
+      await expect(
+        service.delete(reqDetails, {
```
Comment (prestonlimlianjie) [inline services/fileServices/MdPageServices/__tests__/subcollectionPageService.spec.js:182]: Should also expect `mockCollectionYmlService.deleteItemFromOrder` to be called

**Review by prestonlimlianjie — CHANGES_REQUESTED** [review]
Some minor changes but looks great otherwise!

File: services/db/__tests__/gitHubService.js [hunk 3]
```diff
@@ -0,0 +1,559 @@
...
+} = require("@errors/ConflictError")
+const { NotFoundError } = require("@errors/NotFoundError")
+
+const validateStatus = require("@utils/axios-utils")
+
+const { GitHubService } = require("../GitHubService")
+
+const BRANCH_REF = "staging"
+
+describe("Collection Yml Service", () => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply services/db/__tests__/gitHubService.js]: Good catch, resolved in https://github.com/isomerpages/isomercms-backend/pull/259/commits/fd2c3626291e9b012908895bf1fdc94cad57886b

File: services/fileServices/YmlFileServices/__tests__/collectionYmlService.spec.js [hunk 4]
```diff
@@ -0,0 +1,522 @@
...
+  const mockParsedContent = {
+    collections: {
+      [collectionName]: {
+        output: true,
+        order: orderArray,
+      },
+    },
+  }
+  const mockRawContent = yaml.stringify(mockParsedContent)
+  // const mockRawContent = `collections:\n  ${collectionName}:\n    output: true\n    order:\n      - ${fileName}.md\n      - ${subcollectionName}/${subcollectionFileName}.md`
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply services/fileServices/YmlFileServices/__tests__/collectionYmlService.spec.js]: Whoops yeah, deleted in https://github.com/isomerpages/isomercms-backend/pull/259/commits/dcfa84e2f4d260a675f26f2bd500fb05c37a5943

File: services/fileServices/YmlFileServices/__tests__/collectionYmlService.spec.js [hunk 5]
```diff
@@ -0,0 +1,522 @@
...
+      })
+      expect(mockGithubService.update).toHaveBeenCalledWith(reqDetails, {
+        fileName: COLLECTION_FILE_NAME,
+        directoryName,
+        fileContent: modifiedRawContent,
+        sha: oldSha,
+      })
+    })
+
+    it("Adding an third nav page with unspecified index adds it at that position", async () => {
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply services/fileServices/YmlFileServices/__tests__/collectionYmlService.spec.js]: Resolved in https://github.com/isomerpages/isomercms-backend/pull/259/commits/de86fd06354e667ea0f88a0ddc23308ba7cc7613

File: services/fileServices/MdPageServices/__tests__/collectionPageService.spec.js (hunk 1 above)
Comment (alexanderleegs) [reply services/fileServices/MdPageServices/__tests__/collectionPageService.spec.js:170]: Good catch, resolved in https://github.com/isomerpages/isomercms-backend/pull/259/commits/709a7fc18fdc815e314af00330b0e72062b23059

File: services/fileServices/MdPageServices/__tests__/subcollectionPageService.spec.js (hunk 2 above)
Comment (alexanderleegs) [reply services/fileServices/MdPageServices/__tests__/subcollectionPageService.spec.js:182]: Resolved in https://github.com/isomerpages/isomercms-backend/pull/259/commits/709a7fc18fdc815e314af00330b0e72062b23059

File: utils/__tests__/mutex-utils.spec.js [hunk 6]
```diff
@@ -1,4 +1,4 @@
-const { lock, unlock } = require("@utils/mutex-utils")
+const { lock, unlock } = require("../mutex-utils")
```
[the remark this answers is not in the document]
Comment (alexanderleegs) [reply utils/__tests__/mutex-utils.spec.js:1]: I was going for consistency - given that we've moved all tests to a `__tests__` folder within the same directory as the component tested, the imports are always one level up for all classes. Should we change them all to use the alias pathname instead?

[the remark this answers is not in the document]
Comment (alexanderleegs) [reply utils/__tests__/mutex-utils.spec.js:1]: After discussion, we decided to change this to alias pathnames for all tests except for the temporary `newroutes` directory. Resolved in https://github.com/isomerpages/isomercms-backend/pull/259/commits/2d3038f5d077bcb0245d26fb84153990871c5296

**Review by prestonlimlianjie — APPROVED** [review]
lgtm! good work!

**Pull request merged**